### ISSUE TITLE
Annotate EF Core for nullable reference types

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#19007]
~Note: #14150 tracks introducing nullability to non-public parts as well (i.e. annotate the entire codebase), and isn't currently planned for 5.0.~

This issue now tracks annotating the entire codebase for C# 8 NRTs.

/cc @rynowak @cartermp

### Work status

* [x] Analyzers
* [x] Abstractions
* [x] Change tracking (@smitpatel)
* [x] Design (@bricelam)
* [x] Diagnostics
* [x] Extensions
* [x] Infrastructure
* [x] Internal
* [x] Metadata
* [x] Microsoft.Data.Sqlite
* [x] Migrations (@bricelam)
* [x] Proxies
* [x] Query (leftovers in Cosmos, @smitpatel)
* [x] Storage (leftovers in Cosmos, @roji)
* [x] Tools & cmdline (@bricelam)
* [x] Update (@roji)
* [x] ValueGeneration (@roji)

Some specific leftovers we need to do:

* [x] See if we want to somehow pass `nullablePublicOnly` to the compiler - but we use pubternal extensively. Making EF Core more trimming friendly may be an alternative. See https://github.com/dotnet/aspnetcore/pull/29908.
* [x] Take care of MemberIdentity.Name (#23381, @roji). PR: #24437.
* [x] INavigationBase.GetCollectionAccessor (https://github.com/dotnet/efcore/pull/24261/files#r582596243). There's at least one usage which relies on the null-returning behavior, and it's a public API... so not changing anything, at least for now.
* [x] Search remaining instances of `#nullable disable` and see if we can get rid of them
* [x] Search remaining instances of TODO-NULLABLE and see if we can get rid of them
* [x] Value converters (temporarily banged in #23423, proper annotation should be part of #13850, @ajcvickers)

**Comment by ajcvickers** [comment]
Consider specifically the scenario in #19443 when working on this.

**Comment by ajcvickers** [comment]
Tag: C# nullable reference types

**Comment by roji** [comment]
Note: for LINQ async operators (e.g. FirstOrDefaultAsync): https://github.com/dotnet/roslyn/pull/45993 (but depends on C# 9)

**Comment by smitpatel** [comment]
This is done? 
<img src="https://i.redd.it/rwknyu50jmw11.png" />

**Comment by roji** [comment]
> This is done?

Just leave it closed, it will be done in a day or two :rofl: 

Seriously though, any idea why the bot closed this? There wasn't a "closes/fixes" in the commit or PR - is our bot having existential issues?

**Comment by Socolin** [comment]
Hello,

Is there any chance this could end up in the 5.x ? Without having to wait one more year to fix thoses warning without #pragma everywhere in Repository code ?

**Comment by roji** [comment]
@Socolin 5.0 has already been released, and we will not be releasing any new nullability annotations in a 5.0.x patch - so this will only come out in EF 6.0.

However, the good news is that the nullability annotation work is well underway, and so one of the earlier previews should probably already be full annotated - you're very much encouraged to use that when it's out.

**Comment by bricelam** [comment]
This is pretty much a dupe of #14150 now right?

**Comment by roji** [comment]
@bricelam probably... We haven't exactly formally decided we're annotating everything, but that's definitely the way it's going. We can keep this separate for now in case things change later, or close as a dup now.

**Comment by roji** [comment]
@terrajobst you may want to link this from the user story (https://github.com/dotnet/runtime/issues/43619)

**Comment by davidkarlsson** [comment]
Not sure if this has been fixed already but I've ran across a rather strange error that might be easy to miss which I think is related to the ef core analyzer and nullability which causes CS8619 warnings. I reported the issue in the Roslyn repo (https://github.com/dotnet/roslyn/issues/51886) because I didn't see that it was related to EF Core at first.

Basically if you create an anonymous type from a type that has nullable ref types with a lambda expression and you then also call an async ef method in a loop like `FirstOrDefaultAsync()` or `SaveChangesAsync()` in the same method you will get a CS8619 warning saying that the anonymous type doesn't match the expected type.

The same problem happens in both 5.0.4 and 6.0.0-preview.2.21154.2.
